_posts/2025-05-05-date-histogram-optimizations.md

```diff
@@ -0,0 +1,216 @@
+
+```json
+{
+  "size": 0,
+  "aggs": {
+    "dropoffs_over_time ": {
+      "filters": {
+        "1420070400000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-01-01 00:00:00",
+              "lt": "2015-02-01 00:00:00"
+            }
+          }
+        },
+        "1422748800000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-02-01 00:00:00",
+              "lt": "2015-03-01 00:00:00"
+            }
+          }
+        },
+        "1425168000000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-03-01 00:00:00",
+              "lt": "2015-04-01 00:00:00"
+            }
+          }
+        },
+        ...
+      }
+    }
+  }
+}
+```
+
+This date histogram aggregation query generates filter corresponding to each bucket and leverages Lucene’s Points Index, which uses BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and leveraging early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. which are executed in similar manner as range query using the index tree to determine counts for each bucket faster than document value iteration. Similar approach was also applied to the auto date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with consecutive ranges (TODO: Why do we need consecutive ranges??).
+
+![Performance_Improvements_Initial.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/Performance_Improvements_Initial.png)
+
+The following diagrams illustrate how documents are counted per histogram bucket using the index tree. To efficiently count documents matching a range (e.g., 351 to 771), the traversal begins at the root, checking whether the target range intersects with the node’s range. If it does, the algorithm recursively explores the left and right subtrees. An important optimization is skipping entire subtrees: if a node’s range falls completely outside the query range (e.g., 1–200), it is ignored. Conversely, if a node’s range is fully contained within the query range (e.g., 401–600), the algorithm can return the document count from that node directly without traversing its children. This allows the engine to avoid visiting all leaf nodes, focusing only on nodes with partial overlaps. As a result, the operation becomes significantly faster—reducing complexity from linear (O(N)) to logarithmic (O(log N)) in many cases—since it leverages the hierarchical structure to skip large irrelevant portions of the tree and aggregate counts efficiently.
```

Comment:
I don't feel (O(N)) to (O(log N)) is a good way to tell the difference. The real time complexity is much complicated than that. 
I think use the tree to quickly find the target range start, and skip sub trees are good and simple explanation already.

Reply:
Fair point, I have removed all the references to exact time complexity from this blog